identity/src/androidTest/java/com/android/identity/DeviceRetrievalHelperTest.java

```diff
@@ -576,6 +576,182 @@ public void onError(@NonNull Throwable error) {
         Assert.assertTrue(condVarReaderReceivedStatus10.block(5000));
     }
 
+    @Test
+    @SmallTest
+    public void testPresentationBadKeyFromReader() throws Exception {
```

Comment:
Maybe it would be better to make `testPresentation()` a helper function which takes a boolean to determine whether the key is malformed? Then, you would use it as a switch to decide whether to use `encryptMessageToDeviceWithInvalidEReaderKey` as well as whether to use the longer `onDeviceRequest()` in `testPresentation()` or `Assert.fail`.

Though it would add some complexity to the code in `testPresentation()`, it would also reduce the amount of copy-pasted code.

Reply:
Yeah, copy-paste is a problem, there's already three versions of `testPresentation()` and this adds a fourth although the differences are kinda subtle... for test-code, I tend to copy-paste more than to refactor for the simple purpose of ensuring that the test is easy to read, e.g. you can start at the top and not have to jump around to read helper functions or remembering which path to take based on booleans passed in. I've opened Issue #255 for this, would prefer to not address this right now.